### PR TITLE
[feat](Authorization-plugin)Authorization framework modularization

### PR DESCRIPTION
**Pull request by CalvinKirs**
# AccessControllerFactory Interface

## Overview

The `AccessControllerFactory` interface is responsible for creating and managing `CatalogAccessController` instances. The interface includes the following methods:

- **`default String factoryIdentifier()`**: Returns the identifier for the factory, defaulting to the simple name of the implementing class. To maintain compatibility with user-defined plugins from older versions, the `factoryIdentifier` method provides a default implementation that returns the simple name of the current implementation class, ensuring that each factory has a unique identifier.

- **`CatalogAccessController createAccessController(Map<String, String> prop)`**: Creates a new instance of `CatalogAccessController` and initializes it with the provided properties.

## Factory Identifier

Each class implementing `AccessControllerFactory` will automatically use its class name as the factory identifier. This helps in identifying different factory instances during plugin loading and selection.

## Instance Creation

The `createAccessController` method allows you to create and initialize `CatalogAccessController` instances. The `prop` parameter provides the configuration properties needed for initialization.

## Compatibility

- If you are using the previously built-in `range-dorir` authentication plugin, no configuration changes are required; it will continue to function as before.
- For custom plugins, configuration information should be defined in `conf/access.conf`. Then, in `fe.conf`, specify the `access_controller_type` as the identifier for the custom plugin.

## How to Extend

- Add the `fe-core` dependency to your Maven `pom.xml` file.

```xml
        <dependency>
            <groupId>org.apache.doris</groupId>
            <artifactId>fe-core</artifactId>
            <version>1.2-SNAPSHOT</version>
            <scope>provided</scope>
        </dependency>
```
Then, implement the AccessControllerFactory interface to create your own plugin factory class as follows:
```java

public class SimpleAccessControllerFactory implements AccessControllerFactory {
    @Override
    public String factoryIdentifier() {
        return "local-simple";
    }

    @Override
    public CatalogAccessController createAccessController(Map<String, String> map) {
        return new SimpleAccessController(map);
    }
}

package org.example.access;

import org.apache.doris.analysis.ResourceTypeEnum;
import org.apache.doris.analysis.UserIdentity;
import org.apache.doris.common.AuthorizationException;
import org.apache.doris.mysql.privilege.CatalogAccessController;
import org.apache.doris.mysql.privilege.DataMaskPolicy;
import org.apache.doris.mysql.privilege.PrivPredicate;
import org.apache.doris.mysql.privilege.RowFilterPolicy;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

public class SimpleAccessController implements CatalogAccessController {

    private HashMap<String, Boolean> databasePrivs = new HashMap<>();
    // just for test
    public SimpleAccessController(Map<String, String> prop) {
        prop.forEach((k, v) -> {
            databasePrivs.put(k, Boolean.parseBoolean(v));
        });
    }

    @Override
    public boolean checkGlobalPriv(UserIdentity userIdentity, PrivPredicate privPredicate) {
        return false;
    }

    @Override
    public boolean checkCtlPriv(UserIdentity userIdentity, String s, PrivPredicate privPredicate) {
        return true;
    }
    ...

```
Add a new folder named **META-INF/services** under the resources directory, Create a new file named **org.apache.doris.mysql.privilege.AccessControllerFactory.** with a file containing **org.apache.doris.mysql.privilege.AccessControllerFactory.**

## How to Use
- In `fe.conf`, specify the **access_controller_type=local-simple**
- Put the jar file containing the custom plugin in the **fe/custom_lib** directory.

### CODE REVIEW

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run performance

**Review by morningman — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java [hunk 1]
```diff
@@ -52,19 +55,51 @@ public class AccessControllerManager {
...
+            try {
+                prop = PropertiesUtils.loadAccessControllerPropertiesOrNull();
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to load authorization properties,"
+                        + "please check the configuration file, authorization name is " + accessControllerName, e);
+            }
+            return accessControllerFactoriesCache.get(accessControllerName).createAccessController(prop);
+        }
+        throw new RuntimeException("No authorization plugin factory found for " + accessControllerName
+                + "Please confirm that your plugin is placed in the correct location.");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java:91]:
```suggestion
                + ". Please confirm that your plugin is placed in the correct location.");
```

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java [hunk 2]
```diff
@@ -94,23 +129,28 @@ public boolean checkIfAccessControllerExist(String ctl) {
...
-            throw new RuntimeException(e);
-        } catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
+                                       boolean isDryRun) {
+        String pluginIdentifier = getPluginIdentifierForAccessController(acFactoryClassName);
+        CatalogAccessController accessController = accessControllerFactoriesCache.get(pluginIdentifier)
+                .createAccessController(prop);
+        if (!isDryRun) {
+            ctlToCtlAccessController.put(ctl, accessController);
+            LOG.info("create access controller {} for catalog {}", ctl, acFactoryClassName);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java:138]:
```suggestion
            LOG.info("create access controller {} for catalog {}", acFactoryClassName, ctl);
```

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java [hunk 3]
```diff
@@ -52,19 +55,51 @@ public class AccessControllerManager {
...
+        if (accessControllerName.equalsIgnoreCase("default")) {
+            return new InternalAccessController(auth);
+        }
+        if (accessControllerFactoriesCache.containsKey(accessControllerName)) {
+            Map<String, String> prop;
+            try {
+                prop = PropertiesUtils.loadAccessControllerPropertiesOrNull();
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to load authorization properties,"
+                        + "please check the configuration file, authorization name is " + accessControllerName, e);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java:86]:
```suggestion
                        + " Please check the configuration file, authorization name is " + accessControllerName, e);
```

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java [hunk 4]
```diff
@@ -52,19 +55,51 @@ public class AccessControllerManager {
...
+    private CatalogAccessController loadAccessControllerOrThrow(String accessControllerName) {
+        if (accessControllerName.equalsIgnoreCase("default")) {
+            return new InternalAccessController(auth);
+        }
+        if (accessControllerFactoriesCache.containsKey(accessControllerName)) {
+            Map<String, String> prop;
+            try {
+                prop = PropertiesUtils.loadAccessControllerPropertiesOrNull();
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to load authorization properties,"
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/AccessControllerManager.java:85]:
```suggestion
                throw new RuntimeException("Failed to load authorization properties."
```

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Review by zy-kkk — APPROVED** [review]

**Pull request merged**